File: neet_code_all_cpp/math_and_geometry/find_missing_observations.cpp
#include <algorithm>
#include <iostream>
#include <tuple>
#include<vector>

class SolutionAnsKotlin {
    public:
        // AC
        std::vector<int> missingRolls(std::vector<int>& rolls, int mean, int n) {
            int m = rolls.size();
            int m_total = 0; 
            std::for_each(rolls.begin(), rolls.end(), [&](int x) {m_total += x;});
            int n_total = (mean * (m + n)) - m_total;
            if(n_total < n || n_total > n * 6) {
                return std::vector<int>{};
            }

            std::vector<int> result;
            while(n_total > 0) {
                int dice = std::min(n_total - n + 1, 6);
                result.push_back(dice);
                n_total -= dice;
                --n;
            }

            return result;
        }
};

void show_result(std::vector<int>& result) {
    for(auto& v: result) {
        std::cout << v << " ";
    }
    std::cout << '\n';
}

int main(void) {
    using Case = std::tuple<std::vector<int>, int, int>;
    Case case_1 = {{3, 2, 4, 3}, 4, 2};
    // [6, 6}
    Case case_2 = {{1, 5, 6}, 3, 4};
    // [2, 3, 2, 2}
    Case case_3 = {{1, 2, 3, 4}, 6, 4};
    // []

    SolutionAnsKotlin s_ans_kt;

    auto res_1 = s_ans_kt.missingRolls(std::get<0>(case_1), std::get<1>(case_1), std::get<2>(case_1));
    auto res_2 = s_ans_kt.missingRolls(std::get<0>(case_2), std::get<1>(case_2), std::get<2>(case_2));
    auto res_3 = s_ans_kt.missingRolls(std::get<0>(case_3), std::get<1>(case_3), std::get<2>(case_3));

    show_result(res_1);
    show_result(res_2);
    show_result(res_3);
}

File: neet_code_all/src/math_and_geometry/find_missing_observations.rs
// 解けなかった
// 残りのn個の和まではかんたんだが...
struct Solution;
impl Solution {
    pub fn missing_rolls(rolls: Vec<i32>, mean: i32, n: i32) -> Vec<i32> {
        let m = rolls.len() as i32;
        let mut sum_m = 0;
        for r in rolls {
            sum_m += r;
        }

        let sum_n = mean * (m + n) - sum_m;
        println!("sum_n: {sum_n}");

        let mut result = vec![0; n as usize];
        if sum_n % n == 0 {
            for i in 0..n as usize {
                result[i] = sum_n / n;
            }
        } else {
        }

        result
    }
}

// Ktolinの模範解答より
struct SolutionAnsKotlin;
impl SolutionAnsKotlin {
    pub fn missing_rolls(rolls: Vec<i32>, mean: i32, n: i32) -> Vec<i32> {
        let (m, mut n) = (rolls.len() as i32, n);
        let mut n_total = (mean * (n + m)) - rolls.iter().sum::<i32>();

        if n_total < n || n_total > n * 6 {
            return vec![];
        }

        let mut result = vec![];
        while n_total > 0 {
            let dice = std::cmp::min(n_total - n + 1, 6);
            result.push(dice);
            n_total -= dice;
            n -= 1;
        }

        result
    }
}

fn main() {
    let case_1 = (vec![3, 2, 4, 3], 4, 2);
    // [6, 6]
    let case_2 = (vec![1, 5, 6], 3, 4);
    // [2, 3, 2, 2]
    let case_3 = (vec![1, 2, 3, 4], 6, 4);
    // []

    /*
    println!(
        "case_1: {:?}",
        Solution::missing_rolls(case_1.0.clone(), case_1.1, case_1.2)
    );
    println!(
        "case_2: {:?}",
        Solution::missing_rolls(case_2.0.clone(), case_2.1, case_2.2)
    );
    */

    println!(
        "case_1: {:?}",
        SolutionAnsKotlin::missing_rolls(case_1.0.clone(), case_1.1, case_1.2)
    );
    println!(
        "case_2: {:?}",
        SolutionAnsKotlin::missing_rolls(case_2.0.clone(), case_2.1, case_2.2)
    );
    println!(
        "case_3: {:?}",
        SolutionAnsKotlin::missing_rolls(case_3.0.clone(), case_3.1, case_3.2)
    );
}
